pkg/deploy/deploy.go

```diff
@@ -208,51 +210,111 @@ func deployComponentsParallel(ctx context.Context, components []graph.SortedComp
+		}
 	}
 
-	dontDeploy := map[int64]struct{}{} //look-up map marking Nodes (by ID) that should not be deployed, because something they depend on was skipped or failed
+	close(errChan)
+
+	return errs
+}
+
+func (c *componentDeployer) deployNode(ctx context.Context, n graph2.Node) error {
+	cnf := n.(graph.ConfigNode).Config
 
-	for _, gNode := range sortedNodes {
+	ctx = context.WithValue(ctx, log.CtxKeyCoord{}, cnf.Coordinate)
+	entity, err := deployFunc(ctx, cnf, c.clients, c.apis, &c.resolvedEntities)
 
-		node := gNode.(graph.ConfigNode)
-		id := node.ID()
+	// lock changes we will make to shared variables. Writing them is trivial compared to any http request
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if err != nil {
+		failed := !errors.Is(err, skipError)
+		c.removeChildren(ctx, n, n, failed)
 
-		if _, dont := dontDeploy[id]; dont {
-			continue
+		if failed {
+			return err
 		}
+		return nil
+	}
 
-		ctx := context.WithValue(ctx, log.CtxKeyCoord{}, node.Config.Coordinate)
-		entity, err := deployFunc(ctx, node.Config, clientSet, apis, entityMap)
+	c.resolvedEntities.put(*entity)
+	log.WithCtxFields(ctx).Info("Deployment successful")
+	return nil
+}
 
-		if err != nil {
-			deploymentFailed := false
+func (c *componentDeployer) removeChildren(ctx context.Context, parent, root graph2.Node, failed bool) {
 
-			if !errors.Is(err, skipError) {
-				deploymentFailed = true
-				errs = append(errs, err)
-				if !opts.ContinueOnErr && !opts.DryRun {
-					return errs
-				}
-			}
+	parentCnf := parent.(graph.ConfigNode).Config
+	rootCnf := parent.(graph.ConfigNode).Config
 
-			markChildrenAsNotToDeploy(ctx, g, &node, nil, dontDeploy, deploymentFailed)
+	children := c.graph.From(parent.ID())
+	for children.Next() {
+		child := children.Node()
+
+		reason := "was skipped"
+		if failed {
+			reason = "failed to deploy"
+		}
+		childCnf := child.(graph.ConfigNode).Config
 
+		// after the first iteration
+		if parent != root {
+			log.WithCtxFields(ctx).WithFields(field.F("parent", parentCnf.Coordinate), field.F("deploymentFailed", failed), field.F("root", rootCnf.Coordinate)).
+				Warn("Skipping deployment of %v, as it depends on %v which was not deployed after root dependency configuration %v %s", childCnf.Coordinate, parentCnf.Coordinate, rootCnf.Coordinate, reason)
 		} else {
-			entityMap.put(*entity)
-			log.WithCtxFields(ctx).Info("Deployment successful")
+			log.WithCtxFields(ctx).WithFields(field.F("parent", parentCnf.Coordinate), field.F("deploymentFailed", failed)).
+				Warn("Skipping deployment of %v, as it depends on %v which %s", childCnf.Coordinate, parentCnf.Coordinate, reason)
 		}
+
+		c.removeChildren(ctx, child, root, failed)
+
+		c.graph.RemoveNode(child.ID())
 	}
-	return errs
+}
+func deployComponent(ctx context.Context, component graph.SortedComponent, clientSet ClientSet, apis api.APIs, _ DeployConfigsOptions) []error {
+	g := simple.NewDirectedGraph()
+	graph2.Copy(g, component.Graph)
```

Comment:
is this copy necessary?

Reply:
It's a copy to not change the component.
My reasoning is that if you pass the graph in, you might not be expecting it to be completely deleted. My approach is to make a copy so that for the outside the graph does not change at all. Inside, we do remove each and every node of the component.